### ISSUE TITLE
Intermittent failures creating azuredevops_pipeline_authorization 

### ISSUE THREAD

**Issue by ghost** [microsoft/terraform-provider-azuredevops#875]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and Azure DevOps Provider) Version
Terraform v1.5.4
on darwin_arm64
+ provider registry.terraform.io/hashicorp/azuread v2.41.0
+ provider registry.terraform.io/microsoft/azuredevops v0.9.0

### Affected Resource(s)

* `azuredevops_pipeline_authorization`

### Terraform Configuration Files
Partial configuration showing pertinent resources shown below...

```hcl
terraform {
  required_providers {
    azuredevops = {
      source  = "microsoft/azuredevops"
      version = "0.9.0"
    }
    azuread = {
      source  = "hashicorp/azuread"
      version = "2.41.0"
    }
  }
}

locals {
  org_name                 = "mikeguy"
  normalised_project_name  = lower(replace(replace(var.ado_project_name, " ", "-"), "_", "-"))
  environment_names        = nonsensitive(toset([for key, value in var.environments : key]))
  build_service_list       = data.azuredevops_users.build_service.users[*].descriptor
  build_service_descriptor = local.build_service_list[0]
}

resource "azuredevops_project" "this" {
  name               = var.ado_project_name
  visibility         = "private"
  version_control    = "Git"
  work_item_template = "" # Left as blank string to accept org default.
  description        = "Project for ${var.ado_project_name}. Initially holds repositories, variable groups and environments for Terraform configuration."
  features = {
    boards       = "enabled"
    repositories = "enabled"
    pipelines    = "enabled"
    testplans    = "enabled"
    artifacts    = "enabled"
  }
  lifecycle {
    ignore_changes = [work_item_template]
  }
}

resource "azuredevops_git_repository" "this" {
  project_id     = azuredevops_project.this.id
  name           = "terraform-${local.normalised_project_name}-environments"
  default_branch = "refs/heads/main"
  initialization {
    init_type = "Clean"
  }
}

resource "azuredevops_environment" "this" {
  for_each   = local.environment_names
  project_id = azuredevops_project.this.id
  name       = each.value
}

resource "azuredevops_variable_group" "this" {
  for_each     = local.environment_names
  project_id   = azuredevops_project.this.id
  name         = "tfc_variables_${each.value}"
  description  = "Variable group for the ${each.value} environment."
  allow_access = false
  variable {
    name         = "CHECKOV_TOKEN"
    secret_value = var.environments[each.key].checkov_token
    is_secret    = true
  }
  variable {
    name         = "TF_TOKEN"
    secret_value = var.environments[each.key].terraform_token
    is_secret    = true
  }
  variable {
    name  = "DEPLOY_VERSION"
    value = "latest"
  }
  variable {
    name  = "TF_WORKSPACE"
    value = "${var.terraform_workspace_prefix}-${each.key}"
  }
  lifecycle {
    ignore_changes = [variable]
  }
}

resource "azuredevops_pipeline_authorization" "pr_pipeline_environment" {
  for_each    = local.environment_names
  project_id  = azuredevops_project.this.id
  resource_id = azuredevops_environment.this[each.key].id
  type        = "environment"
  pipeline_id = azuredevops_build_definition.pr_pipeline.id
}

resource "azuredevops_pipeline_authorization" "pr_pipeline_vargroup" {
  for_each    = local.environment_names
  project_id  = azuredevops_project.this.id
  resource_id = azuredevops_variable_group.this[each.key].id
  type        = "variablegroup"
  pipeline_id = azuredevops_build_definition.pr_pipeline.id
}
```

### Debug Output
Partial debug output below

```
2023-09-05T10:50:34.438+0100 [TRACE] provider.terraform-provider-azuredevops_v0.9.0: Called downstream: tf_provider_addr=registry.terraform.io/microsoft/azuredevops tf_req_id=ed4c9a2b-77a9-1107-3b42-977d468e1aae tf_resource_type=azuredevops_pipeline_authorization tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.23.0/helper/schema/resource.go:838 @module=sdk.helper_schema timestamp=2023-09-05T10:50:34.437+0100
2023-09-05T10:50:34.439+0100 [TRACE] provider.terraform-provider-azuredevops_v0.9.0: Received downstream response: tf_rpc=ApplyResourceChange tf_req_duration_ms=638 tf_req_id=ed4c9a2b-77a9-1107-3b42-977d468e1aae tf_resource_type=azuredevops_pipeline_authorization diagnostic_error_count=1 diagnostic_warning_count=0 tf_proto_version=5.3 tf_provider_addr=registry.terraform.io/microsoft/azuredevops @caller=github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/internal/tf5serverlogging/downstream_request.go:37 @module=sdk.proto timestamp=2023-09-05T10:50:34.437+0100
2023-09-05T10:50:34.439+0100 [ERROR] provider.terraform-provider-azuredevops_v0.9.0: Response contains error diagnostic: diagnostic_summary=" creating authorized resource: TF400898: An Internal Error Occurred. Activity Id: 11a647ca-7d4c-4536-9467-61832fcd1a9c." tf_provider_addr=registry.terraform.io/microsoft/azuredevops tf_rpc=ApplyResourceChange tf_req_id=ed4c9a2b-77a9-1107-3b42-977d468e1aae @caller=github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/internal/diag/diagnostics.go:55 @module=sdk.proto diagnostic_detail= diagnostic_severity=ERROR tf_proto_version=5.3 tf_resource_type=azuredevops_pipeline_authorization timestamp=2023-09-05T10:50:34.438+0100
2023-09-05T10:50:34.439+0100 [TRACE] maybeTainted: module.ado.azuredevops_pipeline_authorization.pr_pipeline_vargroup["stg"] encountered an error during creation, so it is now marked as tainted
2023-09-05T10:50:34.439+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for module.ado.azuredevops_pipeline_authorization.pr_pipeline_vargroup["stg"]
2023-09-05T10:50:34.439+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: removing state object for module.ado.azuredevops_pipeline_authorization.pr_pipeline_vargroup["stg"]
2023-09-05T10:50:34.439+0100 [TRACE] evalApplyProvisioners: module.ado.azuredevops_pipeline_authorization.pr_pipeline_vargroup["stg"] is tainted, so skipping provisioning
2023-09-05T10:50:34.439+0100 [TRACE] maybeTainted: module.ado.azuredevops_pipeline_authorization.pr_pipeline_vargroup["stg"] was already tainted, so nothing to do
2023-09-05T10:50:34.439+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for module.ado.azuredevops_pipeline_authorization.pr_pipeline_vargroup["stg"]
2023-09-05T10:50:34.439+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: removing state object for module.ado.azuredevops_pipeline_authorization.pr_pipeline_vargroup["stg"]
2023-09-05T10:50:34.439+0100 [TRACE] provider.terraform-provider-azuredevops_v0.9.0: Served request: @module=sdk.proto tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/tf5server/server.go:831 tf_proto_version=5.3 tf_provider_addr=registry.terraform.io/microsoft/azuredevops tf_req_id=ed4c9a2b-77a9-1107-3b42-977d468e1aae tf_resource_type=azuredevops_pipeline_authorization timestamp=2023-09-05T10:50:34.438+0100
2023-09-05T10:50:34.440+0100 [TRACE] statemgr.Filesystem: have already backed up original terraform.tfstate to terraform.tfstate.backup on a previous write
2023-09-05T10:50:34.442+0100 [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 899
2023-09-05T10:50:34.442+0100 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate

```

### Error Output
╷
│ Error:  creating authorized resource: TF400898: An Internal Error Occurred. Activity Id: 166755cf-2ce8-4ebd-92be-39e36fccaf8b.
│ 
│   with module.ado.azuredevops_pipeline_authorization.pr_pipeline_vargroup["tfm"],
│   on module/main.tf line 266, in resource "azuredevops_pipeline_authorization" "pr_pipeline_vargroup":
│  266: resource "azuredevops_pipeline_authorization" "pr_pipeline_vargroup" {
│ 
╵
╷
│ Error:  creating authorized resource: TF400898: An Internal Error Occurred. Activity Id: 11a647ca-7d4c-4536-9467-61832fcd1a9c.
│ 
│   with module.ado.azuredevops_pipeline_authorization.pr_pipeline_vargroup["stg"],
│   on module/main.tf line 266, in resource "azuredevops_pipeline_authorization" "pr_pipeline_vargroup":
│  266: resource "azuredevops_pipeline_authorization" "pr_pipeline_vargroup" {
│ 

### Expected Behavior
When creating the authorization resources with a for_each loop, they should all have been successful. 

### Actual Behavior
Error message above is seen intermittently. After retrying, the number of errors reduces or goes away entirely. Sometimes this will also go through successfully on first run.

I suspect there must be a race condition somewhere or APIs taking time to update on the ADO side. 

### Steps to Reproduce
Run Terraform apply using a fully working configuration. Sometimes it will succeed, sometimes it will fail. This has been seen for both the "environment" types and "variablegroup" types, so don't believe it is specific to one or the other.

**Comment by xuzhang3** [comment]
Provider cannot handle the Internal errors, but we can add a retry logic to retry the failed operation

**Comment by Mabog10** [comment]
We are facing the same issue and it happens more frequently recently..
@xuzhang3 Adding a retry logic might be a good idea, usually rerunning the pipeline helps.

**Comment by xuzhang3** [comment]
@Mabog10 Fixed in #929
